File: sbe/binance/spot_3_1/generated-php/spot_sbe/WebSocketResponse.php
<?php
/**
 * Generated SBE (Simple Binary Encoding) message codec.
 */

class RateLimits
{
    public int|float|null $intervalNum = null;
    public int|float|null $rateLimit = null;
    public int|float|null $current = null;
}

class WebSocketResponse
{
    public const TEMPLATE_ID = 50;
    public const SCHEMA_ID = 3;
    public const SCHEMA_VERSION = 1;
    public const BLOCK_LENGTH = 3;

    public int|float|array|null $sbeSchemaIdVersionDeprecated = null;
    public int|float|array|null $status = null;
    public array $rateLimits = [];
    public string $id = '';
    public string $result = '';

    private function decodeRateLimitsGroup(string $data, int &$offset): array
    {
        $blockLength = unpack('v', substr($data, $offset, 2))[1];
        $offset += 2;
        $numInGroup = unpack('v', substr($data, $offset, 2))[1];
        $offset += 2;

        $items = [];
        for ($i = 0; $i < $numInGroup; $i++) {
            $itemStart = $offset;
            $item = new RateLimits();

            $item->intervalNum = unpack('C', substr($data, $offset, 1))[1];
            $offset += 1;
            $item->rateLimit = unpack('q', substr($data, $offset, 8))[1];
            $offset += 8;
            $item->current = unpack('q', substr($data, $offset, 8))[1];
            $offset += 8;

            // Skip to next block for forward compatibility
            $offset = $itemStart + $blockLength;
            $items[] = $item;
        }

        return $items;
    }

    private function decodeVarData(string $data, int &$offset): string
    {
        $length = unpack('V', substr($data, $offset, 4))[1];
        $offset += 4;
        $value = substr($data, $offset, $length);
        $offset += $length;
        return $value;
    }

    public function encode(): string
    {
        $buffer = '';

        if ($this->status !== null) {
            $buffer .= pack('v', $this->status);
        }

        return $buffer;
    }

    public function decode(string $data): void
    {
        $offset = 0;

        $this->status = unpack('v', substr($data, $offset, 2))[1];
        $offset += 2;

        // Skip to end of block for forward compatibility
        $offset = 3;

        $this->rateLimits = $this->decodeRateLimitsGroup($data, $offset);

        $this->id = $this->decodeVarData($data, $offset);
        $this->result = $this->decodeVarData($data, $offset);
    }
}

File: sbe/binance/spot_3_1/generated-python/spot_sbe/WebSocketResponse.py
"""Generated SBE (Simple Binary Encoding) message codec."""

import struct
from typing import List, Optional, Tuple, Set
from io import BytesIO

class RateLimits:
    """Repeating group item."""

    def __init__(self):
        self.interval_num = None
        self.rate_limit = None
        self.current = None

class WebSocketResponse:
    """SBE message: WebSocketResponse."""

    TEMPLATE_ID = 50
    SCHEMA_ID = 3
    SCHEMA_VERSION = 1
    BLOCK_LENGTH = 3

    def __init__(self):
        self.sbe_schema_id_version_deprecated = None
        self.status = None
        self.rate_limits = []
        self.id = b''
        self.result = b''

    def _decode_rate_limits_group(self, data: bytes, offset: int) -> Tuple[List[RateLimits], int]:
        """Decode repeating group."""
        pos = offset

        block_length = struct.unpack_from('<H', data, pos)[0]
        pos += 2
        num_in_group = struct.unpack_from('<H', data, pos)[0]
        pos += 2

        items = []
        for _ in range(num_in_group):
            item_start = pos
            item = RateLimits()

            item.interval_num = struct.unpack_from('<B', data, pos)[0]
            pos += 1
            item.rate_limit = struct.unpack_from('<q', data, pos)[0]
            pos += 8
            item.current = struct.unpack_from('<q', data, pos)[0]
            pos += 8

            # Skip to next block for forward compatibility
            pos = item_start + block_length
            items.append(item)

        return (items, pos)

    def _decode_var_data(self, data: bytes, offset: int) -> Tuple[bytes, int]:
        """Decode variable-length binary data."""
        pos = offset
        length = struct.unpack_from('<I', data, pos)[0]
        pos += 4
        value = data[pos:pos+length]
        pos += length
        return (value, pos)

    def encode(self) -> bytes:
        """Encode the message to bytes."""
        buffer = BytesIO()

        if self.status is not None:
            buffer.write(struct.pack('<H', self.status))

        return buffer.getvalue()

    def decode(self, data: bytes) -> None:
        """Decode the message from bytes."""
        pos = 0

        self.status = struct.unpack_from('<H', data, pos)[0]
        pos += 2

        # Skip to end of block for forward compatibility
        pos = 3

        self.rate_limits, pos = self._decode_rate_limits_group(data, pos)

        self.id, pos = self._decode_var_data(data, pos)
        self.result, pos = self._decode_var_data(data, pos)
